### PR TITLE
GEODE-8465: secondary HARegionQueue to sync with primary queue

### PR DESCRIPTION
**Pull request by pivotal-eshu**
A seconary HARegionQueue will try to sync with primary after GII
  to remove any events have been already dispatched from primary.

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 1]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
...
+    doReturn(false).when(spy).isRemoved(id1);
+    doReturn(true).when(spy).isRemoved(id2);
+    doReturn(true).when(spy).isRemoved(id3);
+    doReturn(false).when(spy).isRemoved(id4);
+
+    removedEvents = spy.getDispatchedOrRemovedEvents(eventIDS);
+
+    assertThat(removedEvents.size()).isEqualTo(2);
+    assertThat(removedEvents.contains(id2));
+    assertThat(removedEvents.contains(id3));
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:221]: These three assertions can be simplified to `assertThat(removedEvents).containsExactlyInAnyOrder(id2, id3);`

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 2]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
 
     verify(spy).waitForInitialized(time);
   }
+
+  @Test
+  public void getDispatchedOrRemovedEventsReturnsRemovedEvents() {
+    HARegionQueue spy = spy(haRegionQueue);
+    List<EventID> removedEvents;
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:210]: The declaration and assignment of this variable can be moved to the same line.

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 3]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
...
+
+  private void addEvents() {
+    eventIDS.add(id1);
+    eventIDS.add(id2);
+    eventIDS.add(id3);
+    eventIDS.add(id4);
+  }
+
+  @Test
+  public void isRemovedReturnsTrueIfDispatchedAndCurrentEventsAreRemoved() {
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:232]: If the comments about the naming of `getDispatchedOrRemovedEvents()` and `isRemoved()` are applied, this method name should also be changed, along with the following 3 test methods.

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 4]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
...
+    List<EventID> giiEvents;
+    spy.positionBeforeGII = 1;
+    spy.positionAfterGII = 4;
+    HAEventWrapper wrapper1 = mock(HAEventWrapper.class);
+    HAEventWrapper wrapper2 = mock(HAEventWrapper.class);
+    when(wrapper1.getEventId()).thenReturn(id1);
+    when(wrapper2.getEventId()).thenReturn(id2);
+    Region.Entry entry1 = mock(Region.Entry.class);
+    Region.Entry entry2 = mock(Region.Entry.class);
+    Region.Entry entry3 = mock(Region.Entry.class);
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:339]: The IDE warnings for these lines can be removed by using 
`Region.Entry<Object,Object> entry1 = uncheckedCast(mock(Region.Entry.class));`
`Region.Entry<Object,Object> entry2 = uncheckedCast(mock(Region.Entry.class));`
`Region.Entry<Object,Object> entry3 = uncheckedCast(mock(Region.Entry.class));`

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 5]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
...
+    when(entry1.getValue()).thenReturn(wrapper1);
+    when(entry3.getValue()).thenReturn(wrapper2);
+    when(entry2.getValue()).thenReturn(id3);
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:342]: The numbering of things is a little confusing here. Is there a reason that `entry3` returns `wrapper2`  and `entry2` returns `id3`?

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 6]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
...
+    when(haRegion.getEntry(1L)).thenReturn(entry1);
+    when(haRegion.getEntry(2L)).thenReturn(entry2);
+    when(haRegion.getEntry(3L)).thenReturn(null);
+    when(haRegion.getEntry(4L)).thenReturn(entry3);
+
+    giiEvents = spy.getGIIEvents();
+
+    assertThat(giiEvents.size()).isEqualTo(2);
+    assertThat(giiEvents.contains(id1));
+    assertThat(giiEvents.contains(id2));
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:353]: These three assertions can be simplified to `assertThat(giiEvents).containsExactlyInAnyOrder(id1, id2);`

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java [hunk 7]
```diff
@@ -183,4 +203,347 @@ public void isQueueInitializedWithWaitReturnsFalseIfNotInitializedAfterWait() th
...
+  public void getChunksReturnsEqualSizedChunks() {
+    HARegionQueue spy = spy(haRegionQueue);
+    addEvents();
+    // add more events
+    eventIDS.add(mock(EventID.class));
+    eventIDS.add(mock(EventID.class));
+
+    Collection<List<EventID>> myChunks = spy.getChunks(eventIDS, 2);
+
+    assertThat(myChunks.size()).isEqualTo(3);
```
Comment (DonalEvans) [inline geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java:473]: Could this 3 be replaced with a calculated value, and the 2 for max chunk size extracted to a variable? I think that `eventIDS.size()/maxChunkSize` should be the correct value.

**Review by DonalEvans — CHANGES_REQUESTED** [review]
Just a few wording/formatting changes, looks good otherwise.

File: geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java [hunk 8]
```diff
@@ -3943,4 +3958,165 @@ private static int calculateThreadIdExpiryTime() {
...
+  }
+
+  void runSynchronizationWithPrimary(InternalDistributedMember primary, InternalCache cache) {
+    cache.getDistributionManager().getExecutors().getWaitingThreadPool()
+        .execute(() -> doSynchronizationWithPrimary(primary, cache));
+  }
+
+  synchronized void doSynchronizationWithPrimary(InternalDistributedMember primary,
+      InternalCache cache) {
+    if (hasSynchronizedWithPrimary.get()) {
```
[the remark this answers is not in the document]
Comment (pivotal-eshu) [reply geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java:4007]: synchronizeWithPrimaryInProgress should not be true at this point. This is a synchronized method, any thread has set the hasSynchronizedWithPrimary to true, should set synchronizeWithPrimaryInProgress to false in the finally block.

File: geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java (hunk 5 above)
Comment (pivotal-eshu) [reply geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java]: No particular reason, just want to test non HAEventWrapper in an entry case. To be clear, I change the id3 to an object.

Comment (pivotal-eshu) [reply geode-core/src/test/java/org/apache/geode/internal/cache/ha/HARegionQueueTest.java]: Thanks for all the comments and change recommendations.

**Review by DonalEvans — APPROVED** [review]

**Pull request merged**